Emit a the Scala compilation time for downstream metrics

andyscott opened the pull request:
As it stands there isn't clean path to getting target level metrics from Bazel. The build event protocol currently doesn't have timing information for non-test targets [0]. But that doesn't have to stop us from being the first to get this kind of metric from Scala [1].

This adds an output file, the "statsfile" with a very basic property file inspired syntax. The values can be read easily with basic text manipulation utilities, for example `awk -F '=' '$1 == "property_name" { print $2 }' 'path_to.statsfile'` to read the value for `property_name`.

One downside is that this statsfile is non-deterministic. As long as this file is only used for downstream metrics, it shouldn't be problematic. The long term hope is that timing information can be added to the build event protocol and that this PR can be reverted.

### Sample Usage

At Stripe, we're proxying scala rules with macros of the same name. The macro adds additional metrics related target that use the statsfile output. This looks roughly like:

```python
def scala_binary(
        name,
        srcs = [],
        # ...
):
    upstream_scala_binary(...)
    _add_timed_target(name, srcs)

# Adds an extra target named `{name}-timed` that prints out
# timing information to the console
def _add_timed_target(
        name,
        srcs
):
    cmd = """
# ...
TIME_MS=`awk -F '=' '$$1 == "build_time" {{ print $$2 }}' '{statsfile}'` 
# ...
"""
    cmd = cmd.format(statsfile = "$(location :%s.statsfile)" % name)
    native.genrule(
        name = "%s-timed" % name,
        tools = [":%s.statsfile" % name, "//tools:cloc"],
        srcs = srcs, cmd = cmd,
        ...)
```

We're also using https://github.com/AlDanial/cloc to derive the LOC/sec compilation speed. I'd be happy to share the fuller code once all of the kinds are worked out.

[0] Timing support for BEP asked here https://groups.google.com/forum/#!topic/bazel-discuss/WnWq1b9l8Es, by @johnynek 
[1] See open Scala issue https://github.com/scala/bug/issues/10601

bazel-io commented:
Can one of the admins verify this patch?

gkossakowski commented on src/java/io/bazel/rulesscala/scalac/ScalacProcessor.java:
```diff
@@ -247,6 +248,13 @@ private static void compileScalaSources(CompileOptions ops, String[] scalaSource
+      PrintWriter writer = new PrintWriter(ops.statsfile, "UTF-8");
+      writer.println(Long.toString(stop - start));
+      writer.close();
+    } catch (Throwable ex) {
```
catching throwable is a bad idea because it catches JVM errors
you need to catch IOException, right?

johnynek commented on src/java/io/bazel/rulesscala/scalac/ScalacProcessor.java:
```diff
@@ -247,6 +248,13 @@ private static void compileScalaSources(CompileOptions ops, String[] scalaSource
+      writer.println(Long.toString(stop - start));
+      writer.close();
+    } catch (Throwable ex) {
+      throw new RuntimeException("YOLO", ex);
```
can we have a better message here?

ittaiz commented on src/java/io/bazel/rulesscala/scalac/ScalacProcessor.java:
```diff
@@ -247,6 +248,13 @@ private static void compileScalaSources(CompileOptions ops, String[] scalaSource
     }
 
+    try {
+      PrintWriter writer = new PrintWriter(ops.statsfile, "UTF-8");
```
WDYT about using `java.nio.Files#write` instead of these three lines?

andyscott replied:
Sounds good to me

andyscott commented on src/java/io/bazel/rulesscala/scalac/ScalacProcessor.java:
```diff
@@ -247,6 +248,13 @@ private static void compileScalaSources(CompileOptions ops, String[] scalaSource
+      writer.println(Long.toString(stop - start));
+      writer.close();
+    } catch (Throwable ex) {
+      throw new RuntimeException("YOLO", ex);
```
At the time of writing this I figured that if the file fails to write then there must be a show-stopping issue with the build.
- Should the build continue if we can't write the file?
- YOLO might not cut is as the error message :)

johnynek replied:
could we catch IOException instead?

andyscott replied:
I think so. Do we want to catch and continue the build or do we want to rethrow it with a better message?

johnynek commented on src/java/io/bazel/rulesscala/scalac/ScalacProcessor.java:
```diff
@@ -247,6 +248,13 @@ private static void compileScalaSources(CompileOptions ops, String[] scalaSource
 
+    try {
+      PrintWriter writer = new PrintWriter(ops.statsfile, "UTF-8");
+      writer.println(Long.toString(stop - start));
```
should this be a bit more structured? Maybe some hand crafted json even:
```json
{
  build_time: 123123
}
```
so we could add other examples of this?

andyscott replied:
There's no built in way to process JSON downstream, either from common shell scripts, or via Skylark APIs. What about a simple properties style file and I drop the assumption that the build time appears on a given line?
```
baz=false
build_time= 123123
foo=bar
```
This can be read with the one liner:
```bash
BUILD_TIME=`awk -F '=' '$1 == "build_time" { print $2 }' some_target.statsfile`
```

johnynek replied:
sure. Happy to do the above

johnynek commented:
I'm +1. I'll let @ittaiz take a peek before we merge.

ittaiz commented on src/java/io/bazel/rulesscala/scalac/ScalacProcessor.java:
```diff
@@ -247,6 +247,16 @@ private static void compileScalaSources(CompileOptions ops, String[] scalaSource
+    String[] lines = {
+        "build_time=" + Long.toString(stop - start)
+    };
+    Path statsfilePath = FileSystems.getDefault().getPath(ops.statsfile);
```
Nit- wdyt about Paths.get? More canonical

ittaiz commented on src/java/io/bazel/rulesscala/scalac/ScalacProcessor.java:
```diff
@@ -247,6 +247,16 @@ private static void compileScalaSources(CompileOptions ops, String[] scalaSource
+    };
+    Path statsfilePath = FileSystems.getDefault().getPath(ops.statsfile);
+    try {
+        Files.write(statsfilePath, Arrays.asList(lines));
```
Why do we need an array of we convert it to a list? Maybe just create a list two lines above?

andyscott commented:
@ittaiz I've addressed the two small items and added a test.

re: nitpick-- this is totally welcomed. I haven't written plain Java in a long time and I don't remember/know the best patterns or JDK APIs to get the job done.

ittaiz commented on test_rules_scala.sh:
```diff
@@ -734,6 +734,8 @@ $runner bazel run test:MixJavaScalaLibBinary
 $runner bazel run test:MixJavaScalaSrcjarLibBinary
 $runner bazel run test:ScalaBinary
 $runner bazel run test:ScalaLibBinary
+$runner bazel build test:ScalaBinary_statsfile
```
Are you sure we need these? I might be mistaken but I think that `bazel build //test/...` will also catch it and we're doing it in the beginning

andyscott replied:
Oh, good point!

ittaiz commented:
Thanks! liked the test with the genrule. I definitely need to improve my genrule-fu :)
Other than my question about whether or not we need to touch the test_run.sh I'm good.

ittaiz commented on scala/scala.bzl:
```diff
@@ -96,15 +96,17 @@ def _build_nosrc_jar(ctx, buildijar):
     cmd = """
 rm -f {jar_output}
 {zipper} c {jar_output} @{path}
+touch {statsfile}
```
I'm sorry for being too iterative and I appreciate your patience but I just re-read this and wanted to ask- you're doing this (`touch`) since sometimes (only java sources / no sources at all) scala compilation doesn't happen and bazel complained otherwise? I think there's a better way to solve it by only adding it to the `outs`if we need it but since I don't want to complicate you (even more ;) ) I suggest a comment on why we're `touch`ing here?
I just think that it's not that idiomatic and so might be unclear down the line

andyscott replied:
No problem. Scala targets that have no source files get short circuited and don't run the scalac invoker. Creating the statsfile regardless of the number of source files makes it easier to write rules/macros for generating downstream metrics related targets, IMHO.

ittaiz replied:
Ok, sounds good. 
If you think `touch` is enough for these rules/macros, I'm good.
WDYT about adding a comment to explain why we `touch`?

andyscott replied:
I added a brief comment 🙂. Let me know if you want a more thorough one.

katre commented:
I don't have any details about the build event protocol, so I can't help you any there, unfortunately.

ittaiz commented:
Looks good, thanks!
Test this please (for ci bot)
@johnynek
Travis Mac is stuck. I think that if Bazel Build passes we should merge. Wdyt?

johnynek commented:
fracking travis....

johnynek commented:
Thanks for the PR @andyscott !

ittaiz commented:
Indeed, thanks!

On Wed, Jan 24, 2018 at 7:41 PM P. Oscar Boykin <notifications@github.com>
wrote:

> Thanks for the PR @andyscott <https://github.com/andyscott> !
>
> —
> You are receiving this because you were mentioned.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/bazelbuild/rules_scala/pull/406#issuecomment-360214554>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABUIFwVKQX0I5trtUGvYdttWEp78HkZPks5tN2tFgaJpZM4RqTH8>
> .
>
